File: vscodeExt/1.two-sum.java
import java.util.HashMap;
import java.util.Map;

/*
 * @lc app=leetcode id=1 lang=java
 *
 * [1] Two Sum
 */

// @lc code=start
class Solution {
    // by store the counterpart and find it through hash;
    public int[] twoSum(int[] nums, int target) {
        int[] res = new int[2];
        Map<Integer,Integer> m = new HashMap<>();
        // need to store index;
        for (int i = 0;i<nums.length;i++){
            if(m.containsKey(nums[i])){
                res[0] = i;
                res[1] = m.get(nums[i]);
                break;
            } else {
                m.put(target - nums[i], i);
            }
        }


        return res;
    }

    // by brute force;
    public int[] twoSum_bf(int[] nums, int target) {
        int[] res = new int[2];
        for(int i = 0; i<nums.length; i++){
            for (int j =i ;j<nums.length;j++){
                if(nums[i]+ nums[j] == target){
                    res[0] = i;
                    res[1] = j;
                }
            }
        }
        return res;
    }
}
// @lc code=end

File: vscodeExt/1.two-sum.py
#
# @lc app=leetcode id=1 lang=python3
#
# [1] Two Sum
#

# @lc code=start
# Your runtime beats 5.03 % of python3 submissions
# Your memory usage beats 16.51 % of python3 submissions (14.7 MB)
# 
# 注解
# 1，对有序循环，还可以使用双索引，时空都为O(1)
# 2, 对无序循环，使用dict, 空间O(n)


# 暴力双重循环
class Solution_1:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        l = len(nums)
        for i in range(l):
            j = i+1
            while j < l :
                if nums[i] + nums [j] == target:
                    res = [i, j]
                    return res 
                j += 1  


# dict(key:expected number, value: index of original number)
# KeyError 即找不到key. 
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        l = len(nums)
        d = {}
        for i in range(l):
            if nums[i] in d :
                return [ d[nums[i]], i]
            else :
                d[target - nums[i]] = i



# @lc code=end

